crypto/src/base58/encodable.rs

```diff
@@ -20,17 +18,105 @@ pub trait Base58Encodable {
     {
         let mut buf = Vec::<u8>::new();
         to_writer(&mut buf, self).unwrap();
-        encode(buf).with_check_version(Self::VERSION_BYTE)
+        bs58::encode(buf).with_check_version(Self::VERSION_BYTE)
     }
 
-    fn from_base58<'a, I>(i: I) -> Result<Self, Error>
+    fn to_base58_string(&self) -> String
+    where
+        Self: Sized + Serialize,
+    {
+        self.to_base58().into_string()
+    }
+
+    fn from_base58<'a, I>(i: I) -> Result<Self, bin_prot::error::Error>
     where
         I: AsRef<[u8]>,
         Self: Sized + Deserialize<'a>,
     {
-        let bytes: Vec<u8> = decode(i).with_check(Some(Self::VERSION_BYTE)).into_vec()?;
+        let bytes: Vec<u8> = bs58::decode(i)
+            .with_check(Some(Self::VERSION_BYTE))
+            .into_vec()
+            .map_err(|e| bin_prot::error::Error::Custom {
+                message: format!("{:?}", e),
+            })?;
 
         // skip the first byte as this still contains the version byte
-        Ok(from_reader(&bytes[1..]).unwrap())
+        from_reader(&bytes[1..])
+    }
+}
+
+pub trait Base58EncodableHash {
+    /// This is the only part a custom implementation need provide.
+    /// Should be a constant from the base58_version_bytes.rs file corresponding
+    /// to the type.
+    const VERSION_BYTE: u8;
+    const MINA_VERSION_BYTE_COUNT: usize = 1;
+    const MINA_VERSION_BYTE: u8 = 1;
+    fn to_base58(&self) -> EncodeBuilder<'static, Vec<u8>>
```

Comment:
The return type currently restricts `&self` to `'static`. was it not possible with any `'a`?

Reply:
That's just a delegate of the type of `base58::encode`